src/libraries/Microsoft.PowerFx.Connectors/ODataVisitor.cs

```diff
@@ -0,0 +1,237 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Specialized;
+using System.Globalization;
+using Microsoft.PowerFx.Core.IR;
+using Microsoft.PowerFx.Core.IR.Nodes;
+using Microsoft.PowerFx.Core.IR.Symbols;
+using Microsoft.PowerFx.Core.Texl.Builtins;
+using Microsoft.PowerFx.Interpreter;
+using Microsoft.PowerFx.Types;
+
+namespace Microsoft.PowerFx.Connectors
+{
+    internal class ODataVisitor : IRNodeVisitor<string, DelegationRunContext>
+    {
+        internal static readonly ODataVisitor I = new ODataVisitor();
+
+        public override string Visit(TextLiteralNode node, DelegationRunContext runContext)
+        {
+            return $"'{node.LiteralValue}'";
+        }
+
+        public override string Visit(NumberLiteralNode node, DelegationRunContext runContext)
+        {
+            return node.LiteralValue.ToString(CultureInfo.InvariantCulture);
+        }
+
+        public override string Visit(BooleanLiteralNode node, DelegationRunContext runContext)
+        {
+            return node.LiteralValue.ToString(CultureInfo.InvariantCulture);
+        }
+
+        public override string Visit(ColorLiteralNode node, DelegationRunContext runContext)
+        {
+            throw new NotDelegableException();
+        }
+
+        public override string Visit(RecordNode node, DelegationRunContext runContext)
+        {
+            throw new NotDelegableException();
+        }
+
+        public override string Visit(ErrorNode node, DelegationRunContext runContext)
+        {
+            throw new NotDelegableException();
+        }
+
+        public override string Visit(LazyEvalNode node, DelegationRunContext runContext)
+        {
+            return node.Child.Accept(this, runContext);
+        }
+
+        public override string Visit(CallNode node, DelegationRunContext runContext)
+        {
+            // Special cases for functions supported by OData
+            if (node.Function is NotFunction)
+            {
+                return $"NOT({node.Args[0].Accept(this, runContext)})";
+            }
+
+            if (node.Function is StartsWithFunction)
+            {
+                return $"startsWith({node.Args[0].Accept(this, runContext)}, {node.Args[1].Accept(this, runContext)})";
+            }
+
+            if (node.Function is EndsWithFunction)
+            {
+                return $"endsWith({node.Args[0].Accept(this, runContext)}, {node.Args[1].Accept(this, runContext)})";
+            }
+
+            // Fallback: try to evaluate call node, fail if it tries to access a delegated scope
+            try
+            {
+                FormulaValue result = runContext.EvalAsync(node).Result;
+                return SerializeLiteralValue(result);
+            }
+            catch (KeyNotFoundException)
```

Comment:
>KeyNotFoundException [](http://example.com/codeflow?start=19&length=20)

How do we get here?

Reply:
`KeyNotFoundException` is thrown when there is a variable accessed that is not defined in the current scope...effectively this happens when the PowerFX expression has a non-trivial reference of a would-be variable from the Table scope and we cannot delegate the execution

Reply:
This is something that Fx team should followup on and find a better way.  We shouldn't rely on this specific exception

Reply:
Filed #581 to track this.